### ISSUE TITLE
string with format date-time gives validation erro

### ISSUE THREAD

**Issue by jkoshy83** [byu-oit/openapi-enforcer#29]
i have a property in my swagger,
balanceAsOn:            type: "string"           format: date-time           description: "Balance As Of the date and time. The date is in ISO 8601 date format yyyy-mm-ddThh:mm:ssZ"            example: "2019-05-07T09:39:52Z" 

but i am getting validation error:
"Invalid value at: balanceAsOn Expected a valid date object. Received: "2019-05-10T00:00:00:000z""

**Comment by GaryGSC** [comment]
That error doesn't seem as helpful as it could be, so maybe @Gi60s could look into that.

However, your string looks invalid.

Have you tried `"2019-05-10T00:00:00.000Z"`? I believe it needs to match the dateTime regex:
https://github.com/byu-oit/openapi-enforcer/blob/666d8f344a04199c347b0bea1a0d2840c8aa9fea/src/rx.js#L23
Where the regex expects:
- a decimal instead of a colon before the ms (milliseconds are optional)
- a capital Z at the end

**Comment by Gi60s** [comment]
The error states the problem as `Expected a valid date object. Received: "2019-05-10T00:00:00:000z"`.

In more detail, you've given it a date string, but it was expecting an instance of a JavaScript Date object.

How are you using the OpenAPI Enforcer? If you are trying to validate an Open API document then you'll want to check out the docs here: https://byu-oit.github.io/openapi-enforcer/guide/validate-document

If you are attempting to validate input provided via another client (for example, express parsed the body of an HTTP request) then you need to deserialize the input first. For this solution you may be interested in one of these links:

- https://byu-oit.github.io/openapi-enforcer/api/components/openapi#request
- https://byu-oit.github.io/openapi-enforcer/api/components/schema#deserialize

If you are attempting to create an API that follows your OpenAPI document then you might be interested in the [OpenAPI Enforcer Middleware](https://byu-oit.github.io/openapi-enforcer-middleware/).

Let me know if any of that helps or if your question is something else. 

@GaryGSC Thanks for contributing.

**Comment by jkoshy83** [comment]
I am using openapi-enforcer for schema validation(request,response),one of the response shcema contains below property,

balanceAsOn:
          type: "string"
          format: date-time
          description: "Balance As Of the date and time. The date is in ISO 8601 date format yyyy-mm-ddThh:mm:ssZ"
          example: "2019-05-07T09:39:52Z"

and the real response to validate for this property is  "balanceAsOn": "2019-05-10T00:00:00:000z",

    const [schema,error] = new OpenApiEnforcer.v2_0.Schema(schemaDefinition);

            const err = schema.validate(schemaValue);

Even if I providing this a string of format date-time it calls desialize when it executes above code(do I need an explicit call to deserialize?

When I debugging your code real response value is deserialized correctly but the date string mentioned in the example: "2019-05-07T09:39:52Z" is giving the error.

why it validate example in schmea definition,I need to validate only the response?

**Comment by Gi60s** [comment]
I'm not seeing the error. Here is my code that I'm using to test.

```js
const OpenAPI = require('openapi-enforcer');
const definition = {
    type: "string",
    format: 'date-time',
    description: "Balance As Of the date and time. The date is in ISO 8601 date format yyyy-mm-ddThh:mm:ssZ",
    example: "2019-05-07T09:39:52Z",
};
const [ schema, definitionError ] = new OpenAPI.v2_0.Schema(definition);
const deserialized = schema.deserialize('2019-05-07T09:39:52Z');
const validateError = schema.validate(deserialized.value);

console.log(deserialized.value, deserialized.error, deserialized.warning, definitionError, validateError);
```

I'm not seeing any warning or errors.

**Comment by Gi60s** [comment]
Happy to help if you still need it. Closing this issue for now.

**Comment by jkoshy83** [comment]
Hello,

Couple of our APIs having responses with below date-time values:

"dateTime": **"2019-04-16T12:31:22.955595000Z"**
"dateTime": **"2019-12-11T05:43:40+0530"**

open-api enforcer gives validation errors for both these values ;
 Expected a date-time string of the format YYYY-MM-DDTmm:hh:ss.sssZ

We expect these formats should be accepted as it follows **RFC 3339**

**Comment by MaelPic** [comment]
Hello,
About "dateTime": "2019-12-11T05:43:40+0530"
--> should be "2019-12-11T05:43:40+05:30"
See https://tools.ietf.org/html/rfc3339#section-5.6, it says:
`time-numoffset  = ("+" / "-") time-hour ":" time-minute`

Regarding fractions, it says:
`time-secfrac    = "." 1*DIGIT`
which is not 100% clear to me, but at least it doesn't say that fractions are limited to 3 digits, and I agree we should give some freedom there, as some domains require high level of precision (micro-seconds, nano-seconds...)

**Comment by MaelPic** [comment]
Hi @Gi60s ,
I made a PR to allow N digits instead of strictly 3, could you please take a look?
https://github.com/byu-oit/openapi-enforcer/pull/42
Thanks,
Mael

**Comment by jkoshy83** [comment]
Even though this value "2019-12-11T05:43:40+05:30" matches the date-time regex ,it fails at util.getDateFromValidDateString()

**Comment by MaelPic** [comment]
You are right, even `2019-12-11T05:43:40+01:00` doesn't work.
Function getDateFromValidDateString doesn't read the `+01:00` part and simply considers the date is in UTC.

**Comment by Gi60s** [comment]
Thanks for the PR. I'm not at my work computer at the moment so I haven't been able to test your fixes. Does you fix also include the last two comments dealing with the errors in `util.getDateFromValidDateString()`? I don't see a test for that and that is why I ask.

I should be able to get this PR merged in a few hours when I get back to my work computer and I'll work on pushing it to NPM promptly after that.

Thanks again.

**Comment by jkoshy83** [comment]
created a PR to allows a date-time string with +- offset [#43](https://github.com/byu-oit/openapi-enforcer/pull/43)

**Comment by Gi60s** [comment]
Thank you for the PR's. They've been merged and tested. The changes are published to NPM as version `1.6.4`.

**Comment by MaelPic** [comment]
Thank you @Gi60s ! 👍  
Good job @jkoshy83 💯
